Implement per block message ids for contract level messages emitted

alsrdn opened the pull request:
This depends of https://github.com/casper-network/casper-node/pull/4461 and it replaces the per Transaction log implementation we had there. This addresses the comments from the CEP here: https://github.com/casper-network/ceps/pull/88#discussion_r1410061914 

In order to be able to track the ordering of messages that are emitted by different contracts, we now generate an unique ordering number for each message that is emitted during the execution of a block.
    
The first message that is emitted when a block is executed has `block_index=0`, the second message has `block_index=1` and so on.
We also write a record with the total number of messages that were emitted during the execution of a block. This record is purged from global state if a new message is emitted in a new block.

alsrdn committed:
ee/contract messages: add global per block index for contract messages

In order to be able to track the ordering of messages that are emitted
by different contracts, we now generate an unique ordering number for
each message that is emitted during the execution of a block.

The first message that is emitted when a block is executed has
`block_index=0`, the second message has `block_index=2` and so on.
We also write a record with the total number of messages that were
emitted during the execution of a block. This record is purged from
global state if a new message is emitted in a new block.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

Fraser999 commented on types/src/contract_messages/messages.rs:
```diff
@@ -238,6 +266,16 @@ impl Message {
+
+    /// Returns the checksum of the message.
+    pub fn checksum(&self) -> Result<MessageChecksum, bytesrepr::Error> {
+        let mut payload_bytes = self.message.to_bytes()?;
+        let mut block_index_bytes = self.block_index.to_bytes()?;
+        block_index_bytes.append(&mut payload_bytes);
+
+        let checksum = crypto::blake2b(block_index_bytes);
```
```suggestion
        let input = (&self.block_index, &self.message).to_bytes()?;
        let checksum = crypto::blake2b(input);
```

EdHastingsCasperAssociation replied:
@Fraser999  out of curiosity, why would you prefer to bytes over a tuple here? iirc we generally do the append style for multi value hashups.

Fraser999 replied:
Simpler code (at least IMO) but also fewer allocations 1 vs 2 or even 3 (1 per call to `to_bytes()` and possibly one in `append()`).

alsrdn replied:
Yup, good suggestion, changed in: https://github.com/casper-network/casper-node/pull/4550/commits/4550ba389d5e0871c090f901c4c5ac5bd29648fe

Fraser999 commented on types/src/key.rs:
```diff
@@ -1358,6 +1394,7 @@ impl Distribution<Key> for Standard {
             17 => Key::AddressableEntity(rng.gen()),
             18 => Key::ByteCode(rng.gen()),
             19 => Key::Message(rng.gen()),
+            21 => Key::BlockMessageCount(BlockTime::new(rng.gen())),
```
The `match` line for this needs updated (it currently has `match rng.gen_range(0..=18) {`).

alsrdn replied:
Added in https://github.com/casper-network/casper-node/pull/4550/commits/4550ba389d5e0871c090f901c4c5ac5bd29648fe

Fraser999 commented on types/src/key.rs:
We should add a `Key::BlockMessageCount` to the test collection `const KEYS`.

alsrdn replied:
Added in https://github.com/casper-network/casper-node/pull/4550/commits/4550ba389d5e0871c090f901c4c5ac5bd29648fe

alsrdn committed:
Merge branch 'feat-2.0' into contract_messages_per_block_ids

alsrdn committed:
ee/contract-messages: simplify message checksum calculation

Also add some missing tests for `Key::BlockMessageCount`

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

EdHastingsCasperAssociation commented on types/src/key.rs:
```diff
@@ -221,6 +224,8 @@ pub enum Key {
     /// A `Key` under which a single named key entry is stored.
     NamedKey(NamedKeyAddr),
+    /// A `Key` under which the total number of emitted messages in a block is stored.
+    BlockMessageCount(BlockTime),
```
i have some concerns about the address-ability of this...i'd like to discuss it w/ you in realtime when we can sync up

EdHastingsCasperAssociation reviewed (approved):
The broad strokes LGTM. I want to deep dive into the new key variant's payload but this could merge as is and possibly get re-touched in a later PR if necessary.

alsrdn committed:
Merge branch 'feat-2.0' into contract_messages_per_block_ids

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>

alsrdn committed:
ee/contract_messages: use a singleton key for `BlockMessageCount`

Use a singleton `Key::BlockMessageCount` for tracking message counts for
blocks. The value written under this key is a tuple of (BlockTime, u64).
This simplifies the code since we don't need to prune away old values
since the block time is not part of the key anymore.

Signed-off-by: Alexandru Sardan <alexandru@casperlabs.io>
